File: binaryTrees/4.cpp
#include "../common/tree.h"

using namespace Tree;

bool isSameTree(TreeNode *tree_1, TreeNode *tree_2) {
    if (tree_1 == nullptr && tree_2 == nullptr)
        return true;

    if (tree_1 == nullptr || tree_2 == nullptr)
        return false;

    if (tree_1->val != tree_2->val)
        return false;

    return isSameTree(tree_1->left, tree_2->left) &&
           isSameTree(tree_1->right, tree_2->right);
}

int main() {
    TreeNode *tree_1 = createNode(1);
    tree_1->left = createNode(2);
    tree_1->right = createNode(3);

    TreeNode *tree_2 = createNode(1);
    tree_2->left = createNode(2);
    tree_2->right = createNode(3);

    cout << boolalpha << isSameTree(tree_1, tree_2);

    return 0;
}

File: binaryTrees/4.py
from typing import Optional


class TreeNode:
    def __init__(self, val: int):
        self.val = val
        self.left: TreeNode = None
        self.right: TreeNode = None


def isSameTree(p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
    if p == None and q == None:
        return True

    if p == None or q == None:
        return False

    if p.val != q.val:
        return False

    return isSameTree(p.left, q.left) and isSameTree(p.right, q.right)


if __name__ == "__main__":
    root1 = TreeNode(1)
    root1.left = TreeNode(2)
    root1.right = TreeNode(3)

    root2 = TreeNode(1)
    root2.left = TreeNode(2)
    root2.right = TreeNode(3)

    print(isSameTree(root1, root2))
